packages/components/FocusZone/macos/RCTFocusZone.m

```diff
@@ -308,6 +308,73 @@ - (NSView *)nextViewToFocusForAction:(FocusZoneAction)action
 	return [self nextViewToFocusForCondition:block];
 }
 
+- (NSView *)nextViewToFocusForCircularAction:(FocusZoneAction)action
+{
+	BOOL isAdvance = IsAdvanceWithinZoneAction(action);
+	BOOL isHorizontal = IsHorizontalNavigationWithinZoneAction(action);
+
+	NSView *firstResponder = GetFirstResponder([self window]);
+	NSRect firstResponderRect = [firstResponder convertRect:[firstResponder bounds] toView:self];
+	NSPoint anchorPoint = isHorizontal
+		? NSMakePoint(isAdvance ? 0 : [self bounds].size.width, NSMidY(firstResponderRect))
+		: NSMakePoint(NSMidX(firstResponderRect), isAdvance ? 0 : [self bounds].size.height);
+
+	__block CGFloat closestDistance = CGFLOAT_MAX;
+
+	IsViewLeadingCandidateForNextFocus block = ^BOOL(NSView *candidateView)
+	{
+		BOOL isLeadingCandidate = NO;
+		BOOL skip = NO;
+		NSRect candidateRect = [candidateView convertRect:[candidateView bounds] toView:self];
```

Comment:
Was this code lifted from elsewhere, or is to new code? I didn't expect that we would need a new block and all here

Reply:
This was reworked off of the existing navigation code -- the conditions and checks are different -- for example to handle right arrow when we are already at the rightmost control (previous code to determine the control to focus returns nil in this scenario), we attempt to find the leftmost control in roughly the same "row". If we don't find any, then we look in the other (i.e. vertical) axis, again similar to existing navigation code.

It was written this way because this is how I thought about the algorithm in my head -- we're only going to try to wrap when we actually "run out of things". If we can't still find something in one axis, we go in the other axis. Conveniently it also seems less likely to break an existing scenario that doesn't use circular navigation written this way.